pkg/plugins/policies/meshratelimit/plugin/xds/configurer.go

```diff
@@ -0,0 +1,146 @@
+	if err != nil {
+		return err
+	}
+	listenerConfig := &envoy_extensions_filters_http_local_ratelimit_v3.LocalRateLimit{
+		StatPrefix: "rate_limit",
+	}
+	pbstListener, err := proto.MarshalAnyDeterministic(listenerConfig)
+	if err != nil {
+		return err
+	}
+
+	httpRoutes := func(hcm *envoy_hcm.HttpConnectionManager) error {
+		// gateway has dynamic routes so it shouldn't be changed here
+		for _, vh := range hcm.GetRouteConfig().GetVirtualHosts() {
+			routes := vh.GetRoutes()
+			// when size is larger than 1 it means that old ratelimit is applied
+			if len(routes) > 1 {
+				return nil
+			}
+			for _, route := range routes {
+				c.addRateLimitToRoute(route, rateLimit)
+			}
+		}
+		// we don't configure global rate limiting, just enabling it. Problem with enabling
+		// global rate limiting is a filter order.
+		for _, filter := range hcm.HttpFilters {
+			if filter.Name == "envoy.filters.http.local_ratelimit" {
+				return nil
+			}
+		}
+		// envoy.filters.http.router has to be the last filter
+		filters := []*envoy_hcm.HttpFilter{}
+		for _, filter := range hcm.HttpFilters {
+			if filter.Name == "envoy.filters.http.router" {
+				filters = append(filters,
+					&envoy_hcm.HttpFilter{
+						Name: "envoy.filters.http.local_ratelimit",
+						ConfigType: &envoy_hcm.HttpFilter_TypedConfig{
+							TypedConfig: pbstListener,
+						},
+					})
+			}
+			filters = append(filters, filter)
+		}
+		hcm.HttpFilters = filters
+		return nil
+	}
+	if err := listeners_v3.UpdateHTTPConnectionManager(filterChain, httpRoutes); err != nil && !errors.Is(err, &listeners_v3.UnexpectedFilterConfigTypeError{}) {
+		return err
+	}
+	return nil
+}
+
+func (c *Configurer) configureTcpListener(filterChain *envoy_listener.FilterChain) error {
+	config := &envoy_extensions_filters_network_local_ratelimit_v3.LocalRateLimit{
+		StatPrefix: "tcp_rate_limit",
+		TokenBucket: &envoy_type_v3.TokenBucket{
+			MaxTokens:     c.Tcp.Connections,
+			TokensPerFill: proto.UInt32(c.Tcp.Connections),
+			FillInterval:  proto.Duration(c.Tcp.Interval.Duration),
+		},
+	}
+	pbst, err := proto.MarshalAnyDeterministic(config)
+	if err != nil {
+		return err
+	}
+
+	// ratelimiting should be before others filters, except RBAC - we need to ensure
+	// that RBAC is before
+	filters := []*envoy_listener.Filter{}
+	filters = append(filters, &envoy_listener.Filter{
+		Name: "envoy.extensions.filters.network.local_ratelimit.v3.LocalRateLimit",
+		ConfigType: &envoy_listener.Filter_TypedConfig{
+			TypedConfig: pbst,
+		},
+	})
+	filterChain.Filters = append(filters, filterChain.Filters...)
+	return nil
+}
+
+func (c *Configurer) addRateLimitToRoute(route *envoy_route.Route, rateLimit *anypb.Any) {
+	if route.TypedPerFilterConfig == nil {
+		route.TypedPerFilterConfig = map[string]*anypb.Any{}
+	}
+	if route.TypedPerFilterConfig["envoy.filters.http.local_ratelimit"] != nil {
+		return
+	}
+	route.TypedPerFilterConfig["envoy.filters.http.local_ratelimit"] = rateLimit
```

Comment:
so this means only one rate limit can be applied right? what if someone has two policies that target the same route - this will mean only the first one that was applied will work